cardano-node-socket-emulator/src/Cardano/Node/Socket/Emulator.hs

```diff
@@ -75,37 +71,47 @@ main
           Map.fromList $
             zip (getAddress <$> nscInitialTxWallets) (repeat (CardanoAPI.adaValueOf 1_000_000_000))
     initialState <- initialChainState dist
-    let appState = AppState initialState mempty
     params <- liftIO $ Params.fromNodeServerConfig nodeServerConfig
+    let appState = AppState initialState mempty params
     serverHandler <-
       liftIO $
         Server.runServerNode
           (LM.convertLog ProcessingEmulatorMsg trace)
           nscSocketPath
           nscKeptBlocks
           appState
-          params
     serverState <- liftIO $ newMVar appState
     handleDelayEffect $ delayThread (2 :: Second)
 
-    let ctx =
-          Ctx
-            { serverHandler = serverHandler
-            , serverState = serverState
-            , mockTrace = trace
-            }
-
-    runSlotCoordinator ctx
+    let SlotConfig{scSlotZeroTime, scSlotLength} = nscSlotConfig
+    logInfo $
+      StartingSlotCoordination
+        (posixSecondsToUTCTime $ realToFrac scSlotZeroTime / 1_000)
+        (fromInteger scSlotLength :: Millisecond)
+    void $ liftIO $ forkIO $ slotCoordinator nscSlotConfig serverHandler
 
     logInfo $ StartingCNSEServer $ baseUrlPort nscBaseUrl
-    liftIO $ Warp.runSettings warpSettings $ app trace params serverState
+    liftIO $ Warp.runSettings warpSettings $ app trace serverState
     where
       warpSettings = Warp.defaultSettings & Warp.setPort (baseUrlPort nscBaseUrl) & Warp.setBeforeMainLoop whenStarted
 
-      runSlotCoordinator (Ctx serverHandler _ _) = do
-        let SlotConfig{scSlotZeroTime, scSlotLength} = nscSlotConfig
-        logInfo $
-          StartingSlotCoordination
-            (posixSecondsToUTCTime $ realToFrac scSlotZeroTime / 1000)
-            (fromInteger scSlotLength :: Millisecond)
-        void $ liftIO $ forkIO $ slotCoordinator nscSlotConfig serverHandler
+prettyTrace :: Trace IO CNSEServerLogMsg
+prettyTrace = LM.convertLog (renderStrict . layoutPretty defaultLayoutOptions . pretty) stdoutTrace
+
+startTestnet :: FilePath -> Integer -> NetworkId -> IO ()
```

Comment:
Not sure here, but would it be better to have a `withTestnet` function which makes it clear in the user's code that whatever he runs on testnet should wait for the testnet to be available?

Something like:

```
withTestnet :: FilePath -> Integer -> NetworkId -> (() -> IO ()) -> IO ()
```

What do you think?

Reply:
`startTestnet` returns only when the testnet is available, I thought that would be the easiest solution.

Reply:
With the removal of the webserver this is no longer relevant.